exchanges/okx/okx_test.go

```diff
@@ -6093,6 +6094,46 @@ func TestGenerateSubscriptions(t *testing.T) {
 	testsubs.EqualLists(t, exp, subs)
 }
 
+func TestWsCandles(t *testing.T) {
+	t.Parallel()
+	e := new(Exchange) //nolint:govet // Intentional shadow
+	require.NoError(t, testexch.Setup(e), "Setup must not error")
+
+	err := e.WsConnectBusiness(t.Context())
+	require.NoError(t, err, "WsBusinessConnect must not error")
+
+	err = e.BusinessSubscribe(t.Context(), subscription.List{{Channel: channelCandle1D}})
+	require.ErrorIs(t, err, currency.ErrCurrencyPairEmpty)
+
+	p := currency.Pairs{
+		mainPair,
+		currency.NewPairWithDelimiter("ETH", "USDT", "-"),
+		currency.NewPairWithDelimiter("OKB", "USDT", "-"),
+	}
+
+	for i, ch := range []string{channelCandle1D, channelMarkPriceCandle1M, channelIndexCandle1H} {
+		err := e.BusinessSubscribe(t.Context(), subscription.List{{Channel: ch, Pairs: p[i : i+1]}})
+		require.NoErrorf(t, err, "BusinessSubscribe %s-%s must not error", ch, p[i])
+	}
+
+	var got currency.Pairs
+	assert.Eventually(t, func() bool {
+		select {
+		case a := <-e.Websocket.DataHandler:
+			switch v := a.(type) {
+			case websocket.KlineData:
+				got = got.Add(v.Pair)
+			case []CandlestickMarkPrice:
```

Comment:
[bonus points] scrap `CandlestickMarkPrice` in function `wsHandleMarkPriceCandles` and change it to `websocket.KlineData`

```go
candles := make([]websocket.KlineData, len(m.Data))
	for x := range m.Data {
		candles[x] = websocket.KlineData{
			Pair:       m.Argument.InstrumentID,
			Timestamp:  time.UnixMilli(m.Data[x][0].Int64()),
			OpenPrice:  m.Data[x][1].Float64(),
			HighPrice:  m.Data[x][2].Float64(),
			LowPrice:   m.Data[x][3].Float64(),
			ClosePrice: m.Data[x][4].Float64(),
		}
	}
	e.Websocket.DataHandler <- candles
```

```suggestion
			case []kline.KlineData:
```

Reply:
Respectfully declining bonus mission, because:

> TL;RA: There's a lot of out-of-scope here that we're all aware of. Please review this in isolation and don't bother to highlight OOS.

If I start down that road, I *know* I'm going to end up shaving yaks, because I've taken 2 swings at `websocket.KlineData` already in the past.

Reply:
Can't believe you expected me to read, wow

That's fine by me